lightning/src/ln/channel.rs

```diff
@@ -2045,20 +2045,35 @@ struct CommitmentTxInfoCached {
 }
 
 impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
-	fn check_remote_fee<F: Deref, L: Deref>(fee_estimator: &LowerBoundedFeeEstimator<F>,
-		feerate_per_kw: u32, cur_feerate_per_kw: Option<u32>, logger: &L)
-		-> Result<(), ChannelError> where F::Target: FeeEstimator, L::Target: Logger,
+	fn check_remote_fee<F: Deref, L: Deref>(
+		channel_type: &ChannelTypeFeatures, fee_estimator: &LowerBoundedFeeEstimator<F>,
+		feerate_per_kw: u32, cur_feerate_per_kw: Option<u32>, logger: &L
+	) -> Result<(), ChannelError> where F::Target: FeeEstimator, L::Target: Logger,
 	{
 		// We only bound the fee updates on the upper side to prevent completely absurd feerates,
 		// always accepting up to 25 sat/vByte or 10x our fee estimator's "High Priority" fee.
 		// We generally don't care too much if they set the feerate to something very high, but it
-		// could result in the channel being useless due to everything being dust.
-		let upper_limit = cmp::max(250 * 25,
-			fee_estimator.bounded_sat_per_1000_weight(ConfirmationTarget::HighPriority) as u64 * 10);
-		if feerate_per_kw as u64 > upper_limit {
-			return Err(ChannelError::Close(format!("Peer's feerate much too high. Actual: {}. Our expected upper limit: {}", feerate_per_kw, upper_limit)));
-		}
-		let lower_limit = fee_estimator.bounded_sat_per_1000_weight(ConfirmationTarget::Background);
+		// could result in the channel being useless due to everything being dust. This doesn't
+		// apply to channels supporting anchor outputs since HTLC transactions are pre-signed with a
+		// zero fee, so their fee is no longer considered to determine dust limits.
+		if !channel_type.supports_anchors_zero_fee_htlc_tx() {
+			let upper_limit = cmp::max(250 * 25,
+				fee_estimator.bounded_sat_per_1000_weight(ConfirmationTarget::HighPriority) as u64 * 10);
+			if feerate_per_kw as u64 > upper_limit {
+				return Err(ChannelError::Close(format!("Peer's feerate much too high. Actual: {}. Our expected upper limit: {}", feerate_per_kw, upper_limit)));
+			}
+		}
+
+		// We can afford to use a lower bound with anchors than previously since we can now bump
+		// fees when broadcasting our commitment. However, we must still make sure we meet the
+		// minimum mempool feerate, until package relay, is deployed such that we can ensure the
```

Comment:
I think the document should be more precise its package of v3 transactions that don’t have to meet minimum mempool feerate. 

This is documented here as mempool policy in https://github.com/bitcoin/bitcoin/pull/25038/files#diff-ab8886a2beb65c0bfcf0a0c931400f9d5e7e7ff2fbbc9dda985fd1e0ba9b2145R96

"7. An individual V3 transaction is permitted to be below the mempool min relay feerate, assuming it
    is considered within a [package](./packages.md#Package-Feerate) that meets all feerate requirements.”
    
The non-merged / non-deployed p2p Core code is correct on this too: https://github.com/bitcoin/bitcoin/pull/27742/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R1488

Note we might have different nversion=4 in the future, so good to disentangle mempool policy rules from p2p mechanism.

Reply:
> nversion=3 has anything to do with this

As the current state of package relay / nVersion=3 code and documentation is standing in Bitcoin Core main repository, the mempool min fee exemption only apply to nversion=3 transactions not all _transactions_ received from future ancestor package relay (well the pointed code sounds to apply to all, though not sure if it’s a discrepancy).    

>  its static, we can rewrite then when we get there but we don't need to be that forward-looking with our comments, or documentation. 

Yeah I share the sentiment we cannot be that much forward-looking in term (as least as code / doc is not merged on the Core side), still better documentation ease the maintenance burden on our side in case of Core mempool min fee changes, maybe following comment can be applied

```
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 7379c712..0c7ed207 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -2066,8 +2066,9 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 
                // We can afford to use a lower bound with anchors than previously since we can now bump
                // fees when broadcasting our commitment. However, we must still make sure we meet the
-               // minimum mempool feerate, until package relay is deployed, such that we can ensure the
-               // commitment transaction propagates throughout node mempools on its own.
+               // minimum mempool feerate (worst of the dynamic mempool min fee / mempool min relay tx fee)
+               // until package relay and associated mempool policy rules are deployed, such that we can ensure
+               // the commitment transaction propagates throughout node mempools on its own.
                let lower_limit_conf_target = if channel_type.supports_anchors_zero_fee_htlc_tx() {
                        ConfirmationTarget::MempoolMinimum
                } else {
```

Reply:
> We should get that clarified, but either way we currently don't generate v3 transactions, and doing so would be a major effort in lightning as we have to get all our peers to upgrade to some v3-tx format. Given we can't build v3 transactions today, we shouldn't be documenting anything about them!

Clarifications is under way, if someone is bored there is a +3,962 package relay change opened in Core ready for review, altering fundamentally some of the most critical transaction-relay mechanisms :)

> You have rephrased it, but I don't actually get any more information out of it - can you clarify what you think is missing in the current comment?

a) That we must ensure that transactions must be superior to both min relay tx fee and dynamic min fee (and sadly the former might be above the latter even in time of congestion if you have conservative full-node settings for DoS reasons) and b) dissociating package relay as a p2p mechanism (i.e BIP331) from the associated mempool policy rules (nversion=3). As the maximum package size and number of descendants are tied to the policy rules (a.k.a nversion=3) we might have nversion=4 in the future for less sensitive-flow like dual-funding / splicing.

Anyway here a proposed simplification, still mentioning nversion=3 for future code readers to think about interactions in terms of fee update and mempool policy rules

```
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 0c7ed207..9a2ed050 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -2066,9 +2066,9 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 
                // We can afford to use a lower bound with anchors than previously since we can now bump
                // fees when broadcasting our commitment. However, we must still make sure we meet the
-               // minimum mempool feerate (worst of the dynamic mempool min fee / mempool min relay tx fee)
-               // until package relay and associated mempool policy rules are deployed, such that we can ensure
-               // the commitment transaction propagates throughout node mempools on its own.
+               // minimum mempool feerate until BIP331 package relay and associated nversion=3 mempool policy
+               // rules are widely deployed on the network, such that we can ensure the commitment transaction
+               // propagates throughout node mempools on its own.
                let lower_limit_conf_target = if channel_type.supports_anchors_zero_fee_htlc_tx() {
                        ConfirmationTarget::MempoolMinimum
                } else {
```

Reply:
> "Mempool min fee" always readers to the dynamic one, or whatever is required to get into the mempool :)

Oh boy, I’m not sure than all the code comments and documentation in Core is clear on that :)

> Let's please not leave comments referring to things that aren't final and will change out from under us. We'll end up with comments that make no sense in the future and there's no way we're going to remember to update them.

Well package relay might not be never deployed and we’re already speaking about it, though yeah moved the discussion upstream: https://github.com/bitcoin/bitcoin/pull/27742#discussion_r1264309536